server/api/users.js

```diff
@@ -76,29 +76,39 @@ router.post('users.update', auth(), async ctx => {
 });
 
 router.post('users.s3Upload', auth(), async ctx => {
-  const { filename, kind, size } = ctx.body;
-  ctx.assertPresent(filename, 'filename is required');
-  ctx.assertPresent(kind, 'kind is required');
+  let { name, filename, contentType, kind, size } = ctx.body;
+
+  // backwards compatability
+  name = name || filename;
+  contentType = contentType || kind;
+
+  ctx.assertPresent(name, 'name is required');
+  ctx.assertPresent(contentType, 'contentType is required');
   ctx.assertPresent(size, 'size is required');
 
+  const { user } = ctx.state;
   const s3Key = uuid.v4();
-  const key = `uploads/${ctx.state.user.id}/${s3Key}/${filename}`;
+  const key = `uploads/${user.id}/${s3Key}/${name}`;
   const credential = makeCredential();
   const longDate = format(new Date(), 'YYYYMMDDTHHmmss\\Z');
   const policy = makePolicy(credential, longDate);
   const endpoint = publicS3Endpoint();
   const url = `${endpoint}/${key}`;
 
+  await Attachment.create({
+    name,
+    size,
+    url,
```

Comment:
I suspect we should also store the `path` – but it's a little redundant. Maybe store path + url, and make `name` a custom getter that extracts the last piece of the path?

Reply:
Changed to store `key` instead of `name`.